ls1intum/Artemis
# Integrate a minimal Xcode (SwiftUI) template to Artemis

### Checklist

### Motivation and Context


As a follow-up to my Master thesis, where we introduced a Swift programming template to Artemis, we now want to add a minimal template for Xcode programming exercises. This sample will help us to integrate it within the iPraktikum.

### Description

The template consists of a simple Hello World SwiftUI exercise, where the task is to change the name from `Hello, World` to `Hello, Artemis` and let it be displayed in an iOS app.
Furthermore this PR includes:
- a Unit and UI test for the template
- the template uses the ${packageName} variables to be configurable
- the build plan creates a task which compiles, builds and tests the Xcode project
- the build plan sets a requirement for the build agent to use the iOS 14.5 Simulator (only available at testserver1)
- there are two new projectType enums: Plain and Xcode. To distinguish between a plain Swift exercises (like before) and an Xcode programming exercise
- the online code editor and SCA is disabled for Xcode programming exercises

 
### Steps for Testing


Hint: This PR needs to be tested on testserver1! Only ts1 has the agents which fulfill the requirements. Also Jenkins is not supported yet.
**Crucial: If you are done with your testing, please delete the exercise. Otherwise other branches would fail, because they cannot load the Xcode specific config.**

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise
3.1 Now choose Swift as the programming language and Xcode as the project type
3.2 Verify that the SCA activation is not enabled and that a the online code editor is deactivated
3.3 Now choose Plain as project type, activate SCA and online code editor, then switch back to Xcode to verify that these changes are reverted
4. Complete the setup of the template and create the exercise
5. Verify that the template repo has a score of 0% and that the solution repo has 100%
6. Login as a student and participate in the exercise. Verify that you can not use the online code editor but are able to clone it.
7. For the next step you will need to have a Mac in order to open the Xcode project.
7.1 Another way is to open the file structure of the cloned project and open the relevant .swift files within any editor of your choice, change the content and commit it. 
7.2 Have a look into the solution repo to solve the exercise and verify that the student will receive 100% by solving all tasks.


### Test Coverage




FileService.java: 82%
ProjectType.java: 100%
BambooBuildPlanService.java: 94%
BambooProgrammingLanguageFeatureService.java: 100%
JenkinsBuildPlanService.java: 95%
JenkinsProgrammingLanguageFeatureService.java: 100%
ProgrammingExerciseService.java: 92%
ProgrammingExerciseResource.java: 90%
programming-exercise-update.component.ts: 75%

### Screenshots



### Create the Xcode exercise:
![image](https://user-images.githubusercontent.com/33484318/123223962-a2d42e80-d4d1-11eb-9f13-4d3802c929ac.png)

### New problem statement: 
![image](https://user-images.githubusercontent.com/33484318/123224079-be3f3980-d4d1-11eb-8a20-31dfbf990e6c.png)

### No online code editor is enabled:
![image](https://user-images.githubusercontent.com/33484318/123224136-cbf4bf00-d4d1-11eb-93f0-2ffd9954746b.png)

### Template and Solution repo result:
![image](https://user-images.githubusercontent.com/33484318/123224387-0f4f2d80-d4d2-11eb-88fb-ddaab8520095.png)

### Feedback of failed template repo: 
![image](https://user-images.githubusercontent.com/33484318/123224482-242bc100-d4d2-11eb-82ee-7e9a8f17bb42.png)

### Requirements of Bamboo Build Plan:
![image](https://user-images.githubusercontent.com/33484318/123224656-4faeab80-d4d2-11eb-968f-f0fea271914c.png)

### Solved exercise:
![image](https://user-images.githubusercontent.com/33484318/123226314-d6b05380-d4d3-11eb-892d-a453f77cac9f.png)
